File: src/ledger/ledger.ts
import * as fs from 'fs';
import { getClrunPaths, ensureClrunDirs } from '../utils/paths';
import type { LedgerEvent, LedgerEventType } from '../types';

/**
 * Append a structured event to the ledger log.
 * The ledger is append-only.
 */
export function logEvent(
  event: LedgerEventType,
  projectRoot: string,
  terminalId?: string,
  data?: Record<string, unknown>
): void {
  ensureClrunDirs(projectRoot);
  const paths = getClrunPaths(projectRoot);

  const entry: LedgerEvent = {
    timestamp: new Date().toISOString(),
    event,
    ...(terminalId && { terminal_id: terminalId }),
    ...(data && { data }),
  };

  const line = JSON.stringify(entry) + '\n';
  fs.appendFileSync(paths.eventsLog, line, 'utf-8');
}

/**
 * Read all ledger events.
 */
export function readEvents(projectRoot: string): LedgerEvent[] {
  const paths = getClrunPaths(projectRoot);

  if (!fs.existsSync(paths.eventsLog)) {
    return [];
  }

  const content = fs.readFileSync(paths.eventsLog, 'utf-8').trim();
  if (!content) return [];

  return content
    .split('\n')
    .filter(Boolean)
    .map((line) => {
      try {
        return JSON.parse(line) as LedgerEvent;
      } catch {
        return null;
      }
    })
    .filter((e): e is LedgerEvent => e !== null);
}

/**
 * Read the last N events.
 */
export function readRecentEvents(projectRoot: string, count: number): LedgerEvent[] {
  const events = readEvents(projectRoot);
  return events.slice(-count);
}

File: python/clrun/ledger/ledger.py
"""Append-only event ledger for audit trail."""

from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from clrun.utils.paths import get_clrun_paths, ensure_clrun_dirs


def log_event(
    event: str,
    project_root: str,
    terminal_id: Optional[str] = None,
    data: Optional[Dict[str, Any]] = None,
) -> None:
    ensure_clrun_dirs(project_root)
    paths = get_clrun_paths(project_root)
    entry: Dict[str, Any] = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "event": event,
    }
    if terminal_id:
        entry["terminal_id"] = terminal_id
    if data:
        entry["data"] = data
    with open(paths.events_log, "a", encoding="utf-8") as f:
        f.write(json.dumps(entry) + "\n")


def read_events(project_root: str) -> List[Dict[str, Any]]:
    paths = get_clrun_paths(project_root)
    if not os.path.exists(paths.events_log):
        return []
    with open(paths.events_log, "r", encoding="utf-8") as f:
        content = f.read().strip()
    if not content:
        return []
    events: List[Dict[str, Any]] = []
    for line in content.split("\n"):
        if line.strip():
            try:
                events.append(json.loads(line))
            except Exception:
                pass
    return events
